include/vrv/view.h

```diff
@@ -546,19 +546,19 @@ class View {
     void DrawHorizontalSegmentedLine(
         DeviceContext *dc, int y1, SegmentedLine &line, int width, int dashLength = 0, int gapLength = 0);
     void DrawSmuflCode(
-        DeviceContext *dc, int x, int y, wchar_t code, int staffSize, bool dimin, bool setBBGlyph = false);
+        DeviceContext *dc, int x, int y, char32_t code, int staffSize, bool dimin, bool setBBGlyph = false);
     void DrawThickBezierCurve(
         DeviceContext *dc, Point bezier[4], int thickness, int staffSize, int penWidth, int penStyle = AxSOLID);
     void DrawPartFilledRectangle(DeviceContext *dc, int x1, int y1, int x2, int y2, int fillSection);
-    void DrawTextString(DeviceContext *dc, std::wstring str, TextDrawingParams &params);
-    void DrawDirString(DeviceContext *dc, std::wstring str, TextDrawingParams &params);
-    void DrawDynamString(DeviceContext *dc, std::wstring str, TextDrawingParams &params, Rend *rend);
-    void DrawHarmString(DeviceContext *dc, std::wstring str, TextDrawingParams &params);
-    void DrawSmuflLine(DeviceContext *dc, Point orig, int length, int staffSize, bool dimin, wchar_t fill,
-        wchar_t start = 0, wchar_t end = 0);
-    void DrawSmuflString(DeviceContext *dc, int x, int y, std::wstring s, data_HORIZONTALALIGNMENT alignment,
+    void DrawTextString(DeviceContext *dc, std::u32string str, TextDrawingParams &params);
```

Comment:
Here and the following: Pass strings as const reference whenever possible.

Reply:
Yes, but here that is not possible. I changed it because I had to drop the `std::stringstream` and tokenize the string 'by hand'. This means modifying it. 
https://github.com/rism-digital/verovio/blob/3f15f1f9034edd4ced47d34cc2bc9aaea94fc652/src/view_text.cpp#L291
If we pass it as reference we need to make a copy, which would not make it more efficient.

Reply:
Yes, this is in `DrawLyricString` and I agree that passing by value is better there. For the other functions:

- `DrawTextString`: `str` is not modified and hence should be passed as const ref.
- `DrawDirString`: `str` is not modified, but copied at the first opportunity into `adjustedStr` which then is modified. Here we should keep passing `str` by value and remove the local variable `adjustedStr`.
-  `DrawDynamString`:  `str` is not modified and hence should be passed as const ref.
-  `DrawHarmString`:  `str` is not modified and hence should be passed as const ref.

Reply:
In such cases I pass by const reference, and then copy the string inside of the function.  In such cases there will be no loss in efficiency since there has to be a copy made somewhere, so I am doing it explicitly rather than implicitly.   This makes the function declaration look nicer and you don not have to wonder if the const reference was not added by accident.

Reply:
Yes, this is probably clearer. I'll change it.

Reply:
> In such cases there will be no loss in efficiency since there has to be a copy made somewhere

Just want to remark that this is not true if an rvalue is passed into the function. If the string is passed by value, then the string will be moved and not copied in this case which is more efficient. It’s like copying a pointer instead of a whole container.
However, let’s leave it as above and passing everything by const ref should be good enough, I’m happy with that.